ENH: float16.h in NumPy

athurdekoos opened the pull request:
This issue will close #30280

## Summary of Changes:
- New `float16.h` 
  - This is a header only implementation of `halffloat.h`. Effectively eliminating the need to include `npy_math` when using `halffloat.h` functions. 
  - A majority of the logic has been `Tactically Acquired` from the old [`halffloat.c`](https://github.com/numpy/numpy/blob/bba99dbd606d173a12c7de1867e2337de17c60e9/numpy/core/src/npymath/halffloat.c) implementation to reduce the need to rewrite everything form scratch. 
  - I have placed header guards in `float16.h` where needed to allow for both `halffloat.h` and `float16.h` in case of simultaneous includes. Functions have also been renamed to avoid conflict.
  - `#define NPY_HALF_GENERATE_OVERFLOW ` and `#define NPY_HALF_GENERATE_INVALID` has been removed and logic has been updated to remove the code path.
 
- New `test_float16.py` and `_float16_tests.c` to support tests for float16

Update 12/8/25:
- New `tests/examples/float16/` directory has been created.
  - `_float16_tests.c` has been relocated to this directory.
  - New `float16/setup.py` & `float16/meson.build` files have been created to assist with testing. 
- Update in`test_float16.py` with logic similar to [tests/test_limited_api.py](https://github.com/numpy/numpy/blob/60bf05c38a1059cef0d897a83c626bad702ae1ba/numpy/_core/tests/test_limited_api.py) to better test `float16.h` per [Ralf's recommendation ](https://github.com/numpy/numpy/pull/30380#pullrequestreview-3549564449) (thanks Ralf!)

Update 12/9/25:
- New `float16.rst` document has been added for documentation on the new `float16.h` file. 
- Update in `coremath.rst`, a user facing note has been added to steer users to `float16.h`.
- Update in`c-api/index.rst` to add the new `float16.rst` document to the file tree.

Update 12/10/25:
- Update in `coremath.rst` did not have a a reference to `NPY_MAX_HALF`, a new one has been created and added. 
- Update in `float16.h`,  `npy_float16_spacing` was removed.
  -  `_npy_float16_header_invalid` was removed as it is now unneeded.

Commit:
WIP: float16.h in NumPy

WIP: Verifying new testing procedure in examples/

WIP: Updates for linting, adding IS_EDITABLE and IS_WASM

WIP: Linting update in tests/test_float16.py

athurdekoos commented:
I would love some feedback on[ _float16_tests.c ](https://github.com/numpy/numpy/blob/ff6ada95a9dac0457d9b86dcfa4510a60c2e9c23/numpy/_core/src/npymath/_float16_tests.c)

I have the same function logic essentially repeated multiple times right now to help with debug but I am a bit back and forth on if I should just factor this out.

rgommers reviewed (commented):
Thanks @athurdekoos. The implementation looks pretty solid. I didn't verify line by line that it matches the old `halffloat.c` content yet, but I assume it does - and everything else LGTM. The one thing there is to decide whether to include all functions or not, which can be left to the very end.

The tests look less ideal to me. The compilation uses `common.h` which wouldn't work for an external user. And shipping a pre-compiled extension module increases the binary size and is less good of a test than compiling something at test time. That's what is done for some other tests, like in `test_cython.py` and `test_limited_api.py`. I think the limited API one is the example to follow:
- The C code that uses `float16.h` goes under `_core/tests/examples/` (like https://github.com/numpy/numpy/tree/60bf05c38a1059cef0d897a83c626bad702ae1ba/numpy/_core/tests/examples/limited_api)
- The `test_float16.py` can use this pattern to compile an extension module at test time: https://github.com/numpy/numpy/blob/60bf05c38a1059cef0d897a83c626bad702ae1ba/numpy/_core/tests/test_limited_api.py
    - Note that the test skips will be simpler than for the limited API test - you should only need the `IS_EDITABLE` and `IS_WASM` ones

Could you try to change to that testing method?

Commit:
ENH: float16.h in NumPy

rgommers commented:
Looks good now, I marked it as ready for review.

rgommers commented:
Actually modulo docs that it. The docs should explain the `float16.h` usage. And I think https://numpy.org/devdocs/reference/c-api/coremath.html#half-precision-functions should get a `.. note::` recommending to not use `halffloat.h` anymore but `float16.h` instead because it doesn't require linking `libnpymath`.

Commit:
ENH: float16.h documentation and comments in NumPy.

ENH: float16.h in NumPy, documentation formating update

ENH: float16.h in NumPy, documentation formating update

athurdekoos commented:
`c-api/float16.rst` was added as new stand alone file and added to the file tree under c-api. I recognize that this is likely not the most ideal location for this document, however I also did not believe it should be included in `c-api/coremath.rst`. 

Feedback on appropriate location for this document would be appreciated.

athurdekoos commented on numpy/_core/tests/examples/float16/meson.build:
```diff
@@ -0,0 +1,20 @@
+  include_directories: [npy_include_path],
+    c_args: [
+      '-DNPY_TARGET_VERSION=NPY_2_0_API_VERSION',
+    ],
```
Removed

rgommers replied:
note that indentation is still off in this file, it should be consistently 2 or 4 spaces. You can run `meson fmt` to autofix

rgommers commented on doc/source/reference/c-api/float16.rst:
```diff
@@ -0,0 +1,263 @@
+Constants
+---------
+
+:c:macro:: NPY_HALF_ZERO
```
These aren't rendered correctly, see the rendered version in CI

Commit:
ENH: sphinix formating fix in float16.rst

ENH: sphinix formating fix in float16.rst

ENH: sphinix formating fix in float16.rst

Commit:
ENH: Removing npy_float16_spacing and ref for float16.h

ENH: Removing npy_float16_spacing and ref for float16.h

seberg commented on doc/source/reference/c-api/float16.rst:
```diff
@@ -0,0 +1,248 @@
+==================
+
+The header file ``<numpy/float16.h>`` provides helper routines for working
+with IEEE 754-2008 16-bit floating-point values
```
Not sure where, but the `float16.h` header is special in the sense that none of the functions here require "importing" NumPy (since they don't call any functions).
I think it would be nice to mention that somewhere.

I guess we need `ndarraytypes.h` to have `npy_float16` and `npy_half` defined? Would be cool if we didn't need the full thing, or need it at all (it's rather large).
But I am not really worried about that.

seberg commented on numpy/_core/tests/examples/float16/_float16_tests.c:
```diff
@@ -0,0 +1,343 @@
+    if (PyModule_AddObject(m, "float16", m) < 0) {
+        Py_DECREF(m);
+        return -1;
+    }
```
Why would the module add itself to the namespace?

seberg commented on numpy/_core/tests/examples/float16/_float16_tests.c:
```diff
@@ -0,0 +1,343 @@
+    npy_half mod;
+    npy_half quo = npy_float16_divmod(h1, h2, &mod);
+
+    return Py_BuildValue("II", quo, mod);
```
This seems incorrect on some platforms. You need to make sure the types match. I assume `I` is `unsigned int`, so you have to type or cast `mod` and `quo` accordingly.  (`PyLong_FromUnsignedLong` is OK, since it is a typed function.)
(I have not checked if there are more occurrences of this)

The same is true for `ParseTuple`, but I think there it is correct everywhere.

seberg reviewed (commented):
A few comments, but overall looks good to me, thanks.  I am not quite sure we have settled if we want to remove anything again here.

athurdekoos commented:
> A few comments, but overall looks good to me, thanks. I am not quite sure we have settled if we want to remove anything again here.

I think it's worth removing the *nonane variants. Is there anything else that should go?

Commit:
ENH: Removed divmod & *Nonane in float16.h

float16.h
- Removed math.h as it is no longer needed
- Updated constants to new float16 naming convention and removed header
guards
- Removed npy_float16_eq_nonan, divmod, and _divemodf functions
- Updated npy_float16_lt_nonane and npy_float16_le_nonan private
functions _npy_float16_lt_nonane and _npy_float16_le_nonan

halffloat.h
- Added depreciation warning comment

_float16_tests.c
- Removed float16_eq _nonane wrapper
- Updated npy_float16_lt_nonane and npy_float16_le_nonan wrappers to
_npy_float16_lt_nonane and _npy_float16_le_nonan
- Removed divmod wrapper
- Removed float16tests_exec and float16tests_slots[]
- Simplified implementation per Sebastians review with freethreading
implementation

test_floats16.py
- Removed float16_eq_nonane tests
- Updated npy_float16_lt_nonane and npy_float16_le_nonan to
_npy_float16_lt_nonane and _npy_float16_le_nonan
- Removed dived tests

float16.rst
-Updated wording to reflect that NumPy C-API does not need to imported
via import_array()
- Constants have been updated to reflect new float16 naming.
- Constants have been updated to macros from refrences
- Removal of Nan section and Divmod section to reflect updates in
float16.h

30380.c_api.rst
- Created release note fragment

seberg commented:
Thanks for the update. From my side, I happy if this gets in as is (and if there is feedback we can incorporate it).
If we end up with the deprecation warning might as well wait till after 2.4 is released (around the corner anyway).

This should get a release note fragment as well. It can be very brief though and mostly link to the other (very nice!) docs.  (If you have not done that yet, see `doc/release/upcoming_changes/README.md`)

athurdekoos committed:
ENH: Updated formating

Removed extraneous comment
